core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2319,125 +2324,117 @@ class KafkaController(val config: KafkaConfig,
-                new AlterPartitionResponseData.PartitionData()
-                  .setPartitionIndex(tp.partition)
-                  .setLeaderId(leaderAndIsr.leader)
-                  .setLeaderEpoch(leaderAndIsr.leaderEpoch)
-                  .setIsr(leaderAndIsr.isr.map(Integer.valueOf).asJava)
-                  .setLeaderRecoveryState(leaderAndIsr.leaderRecoveryState.value)
-                  .setPartitionEpoch(leaderAndIsr.partitionEpoch)
-              )
-          }
+    // Update our cache and send out metadata updates
+    updateLeaderAndIsrCache(successfulUpdates.keys.toSeq)
+    sendUpdateMetadataRequest(
+      controllerContext.liveOrShuttingDownBrokerIds.toSeq,
+      partitionsToAlter.keySet
+    )
+
+    partitionResponses.groupBy(_._1.topic).forKeyValue { (topicName, partitionResponses) =>
+      // Add each topic part to the response
+      val topicResponse = if (useTopicsIds) {
+        new AlterPartitionResponseData.TopicData()
+          .setTopicId(controllerContext.topicIds.getOrElse(topicName, Uuid.ZERO_UUID))
+      } else {
+        new AlterPartitionResponseData.TopicData()
+          .setTopicName(topicName)
+      }
+      alterPartitionResponse.topics.add(topicResponse)
+
+      partitionResponses.forKeyValue { (tp, errorOrIsr) =>
+        // Add each partition part to the response (new ISR or error)
+        errorOrIsr match {
+          case Left(error) =>
+            topicResponse.partitions.add(
+              new AlterPartitionResponseData.PartitionData()
+                .setPartitionIndex(tp.partition)
+                .setErrorCode(error.code))
+          case Right(leaderAndIsr) =>
+            /* Setting the LeaderRecoveryState field is always safe because it will always be the same
+             * as the value set in the request. For version 0, that is always the default RECOVERED
+             * which is ignored when serializing to version 0. For any other version, the
+             * LeaderRecoveryState field is supported.
+             */
+            topicResponse.partitions.add(
+              new AlterPartitionResponseData.PartitionData()
+                .setPartitionIndex(tp.partition)
+                .setLeaderId(leaderAndIsr.leader)
+                .setLeaderEpoch(leaderAndIsr.leaderEpoch)
+                .setIsr(leaderAndIsr.isr.map(Integer.valueOf).asJava)
+                .setLeaderRecoveryState(leaderAndIsr.leaderRecoveryState.value)
+                .setPartitionEpoch(leaderAndIsr.partitionEpoch)
+            )
         }
       }
-
-      callback(alterPartitionResponse)
-    } catch {
-      case e: Throwable =>
-        error(s"Error when processing AlterPartition for partitions: ${partitionsToAlter.keys.toSeq}", e)
-        callback(new AlterPartitionResponseData().setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code))
-        // We consider that all partitions have failed.
-        partitionResponses.clear()
     }
 
+    callback(alterPartitionResponse)
```

Comment:
Is it possible for the logic below to raise an exception? Would there be any consequences to invoking the callback twice?

Reply:
It is hard to say because `maybeCompleteReassignment` call stack is pretty deep. I have reworked this part to run the below code outside of the try...catch. This ensure that the callback is called only once and only for exceptions non related to the below code.